thinc/loss.py

```diff
@@ -67,10 +71,30 @@ def _validate_input(self, guesses: FloatsOrRaggedT, target: Floats2d) -> None:
                 "with truth values outside the [0,1] interval."
             )
 
+    def _get_sample_weights(self, target: Floats2d) -> Floats1d:
```

Comment:
I think it makes sense but not sure how well it composes, because I think it would require a fine-grained type system to actually make it fool proof. What I'm thinking is that let's say we implement a `ClassWeighter` that takes a loss as input, but then this loss has to have type `CategoricalLoss`, which could be the cross-entropy or KL or some margin-based etc. Not only that, but it also doesn't compose with `Sequential` type losses only with the internal loss of the `Sequential` loss. 

So we I can go ahead and implement this `ClassWeighter` outside of the cross-entropy losses, but it will just factor out the code for validating and converting the `class_weights` + adjusting the gradients, but I don't think we can make it into this nice general thing.

Reply:
> What would work better maybe is to use a CrossEntropyWeighter for now that can only take a loss that is isinstance of CategoricalCrossentropyBase and then do the work?

👍 

Or perhaps even just `CategoricalCrossentropy` and a separate weighter for the sequence case. They could share code.

Reply:
Oh wait, a problem with the wrapper class is that it will have the wrong type right? I can wrap the `CategoricalCrossentropy` in `CrossEntropyWeighter`, but then `CrossEntropyWeighter` has to be also of type `CategoricalCrossEntropyBase` right?

Reply:
I've been looking into it today and it doesn't seem to compose too well without refactoring the cross-entropy losses again. The weighting comes in right at the gradient calculation in `_get_grad`

```python
differrence = _to_array(guesses) - target
do weighting here
do masking
do normalization
```
Also in `_get_loss` the sample weights need to be injected before summing the loss terms and normalization. So I think I would need to re-write the API yet again to compose well right?